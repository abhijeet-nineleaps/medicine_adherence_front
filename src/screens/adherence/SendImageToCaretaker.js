import {
  Image,
  Modal,
  ScrollView,
  TouchableOpacity,
  View,
  LogBox,
} from 'react-native';
import {API_URL} from '../../repositories/var';
import React, {useState, useEffect} from 'react';
import BouncyCheckbox from 'react-native-bouncy-checkbox';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {useFocusEffect} from '@react-navigation/native';
import {Button, Text} from 'react-native-elements';
import * as Progress from 'react-native-progress';
import {useRoute} from '@react-navigation/native';
import Toast from 'react-native-toast-message';
import Share from 'react-native-share';
import queryData from '../../repositories/database/queryData';
import globalDb from '../../repositories/database/globalDb';
import styles from './adherenceStyles/SendImageToCareTakerStyles';
import {useDispatch} from 'react-redux';
import fetchCaretakers from '../../redux/actions/caretaker/CaretakerActions';
import Logger from '../../components/logger';
LogBox.ignoreLogs(['Require cycle:']);
const db = globalDb();
let medName = '';
let medId = 0;
const SendImageToCaretaker = ({navigation}) => {
  const route = useRoute();
  const image_uri = route.params;
  // const {image_uri} = route.params;
  const [mycaretakers, mycaretakerstate] = useState([]);
  const [send_to, send_to_state] = useState('');
  const [modalVisible, setModalVisible] = useState(false);
  const [medsArray, medsArrayState] = useState([]);

  const Renderitem = ({item}) => {
    const [med1, setMed1] = useState(false);
    const setMedFnc = () => {
      setMed1(!med1);
      !med1 ? send_to_state(item.caretakerId) : send_to_state('');
    };
    return (
      <View style={styles.cnConatiner}>
        <BouncyCheckbox
          size={22}
          fillColor="#3743ab"
          unfillColor="#FFFFFF"
          isChecked={med1}
          iconStyle={styles.cnChechBoxIcon}
          textStyle={styles.cnCheckBoxText}
          disableBuiltInState
          id="setMed"
          onPress={() => setMedFnc}
        />
        <Text style={styles.cnText}>{item.caretakerUsername}</Text>
      </View>
    );
  };

  const RenderMeds = ({item}) => {
    const [clicked, clickedstate] = useState(false);
    const [borderColo, borderColorstate] = useState('white');
    const [tcolor, tcolorstate] = useState('black');
    const clickedFnc = () => {
      medName = item.medicine_name;
      medId = item.user_id;
      clickedstate(!clicked);
      if (clicked) {
        borderColorstate('white');
      } else {
        borderColorstate('#4dd0e1');
        tcolorstate('white');
      }
    };
    return (
      <View
        style={{
          padding: 10,
          margin: 10,
          borderColor: 'black',
          borderWidth: 0.1,
          height: 40,
          alignItems: 'center',
          justifyContent: 'center',
          backgroundColor: borderColo,
          borderRadius: 20,
        }}>
        <TouchableOpacity id="clicked" onPress={() => clickedFnc}>
          <Text style={{fontWeight: '500', fontSize: 15, color: tcolor}}>
            {item.medicine_name}
          </Text>
        </TouchableOpacity>
      </View>
    );
  };

  const [_refresh, refeereshstate] = React.useState(false);
  const dispatch = useDispatch();
  const fetchcaretakers = async () => {
    let user_id = await AsyncStorage.getItem('user_id');
    dispatch(fetchCaretakers(user_id));
    refeereshstate(false);
  };

  useEffect(() => {
    fetchcaretakers();
  }, []);
  const renderitem = ({item}) => {
    Logger.loggerInfo(item.patientId);

    const fetchMedicines = async () => {
      db.transaction(async txn => {
        let medsArr = await queryData.getusermeds(txn);
        medsArrayState(medsArr);
      });
    };
    useFocusEffect(
      React.useCallback(() => {
        async function name() {
          await fetchMedicines();
          let value = await fetchcaretakers();
          mycaretakerstate(value);
        }
        name();
        return () => {
          /*do nothing*/
        };
      }, []),
    );
  };
  async function SendImage() {
    setModalVisible(true);
    if (medName === '') {
      setModalVisible(false);
      Toast.show({
        type: 'info',
        text1: 'Select medicine',
      });
      return;
    }
    let todayDate = new Date();
    let setDate =
      todayDate.getDate() +
      '-' +
      (todayDate.getMonth() + 1) +
      '-' +
      todayDate.getFullYear();

    let imagesData = await AsyncStorage.getItem(setDate + ' ' + medName);
    if (imagesData !== null) {
      let parsedData = JSON.parse(imagesData);
      parsedData.push(image_uri);
      await AsyncStorage.setItem(
        setDate + ' ' + medName,
        JSON.stringify(parsedData),
      );
    } else {
      let parsedData = [];
      parsedData.push(image_uri);

      await AsyncStorage.setItem(
        setDate + ' ' + medName,
        JSON.stringify(parsedData),
      );
    }
    const formdata = new FormData();
    var dt = new Date().getTime();

    let patientName = await AsyncStorage.getItem('user_name');

    var file_name = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(
      /[xy]/g,
      function (c) {
        var r = (dt + Math.random() * 16) % 16 | 0;
        dt = Math.floor(dt / 16);
        return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
      },
    );
    formdata.append('image', {
      name: 'care',
      uri: image_uri,
      type: 'image/jpg',
    });
    formdata.append('name', file_name);
    formdata.append('id', send_to);
    formdata.append('medName', medName + ' taken by ' + patientName);
    formdata.append('medId', medId);
    const url = `${API_URL}`;
    fetch(url + '/api/v1/image', {
      method: 'post',
      headers: {
        'Content-Type': 'multipart/form-data',
      },
      body: formdata,
    })
      .then(() => {
        setModalVisible(false);
        Toast.show({
          type: 'success',
          text1: 'Image sent',
        });
        setTimeout(() => {
          navigation.pop(1);
        }, 1000);
      })
      .catch(() => {
        setModalVisible(false);
      });
  }
  async function shareFnc() {
    const shareOptions = {
      title: 'Share file',
      email: 'email@example.com',
      social: Share.Social.EMAIL,
      failOnCancel: false,
      urls: [image_uri],
    };
    await Share.open(shareOptions);
  }
  return (
    <View style={styles.container}>
      <Toast visibilityTime={1500}></Toast>
      <Modal
        animationType="fade"
        transparent={true}
        visible={modalVisible}
        style={styles.modal}>
        <View style={styles.modalOuterView}>
          <View style={styles.modalInnerView}>
            <Text>Please wait Uploading Image!</Text>
            <Progress.CircleSnail
              spinDuration={500}
              size={80}
              color={['red', 'green', 'yellow']}
            />
          </View>
        </View>
      </Modal>
      <View style={styles.container1}>
        <Text style={styles.container1Text}>Image</Text>
        <Button title="Share" id="share" onPress={shareFnc}></Button>
      </View>
      <Image source={{uri: image_uri}} style={styles.image}></Image>
      <ScrollView>
        <View style={styles.mnView}>
          <Text style={styles.mnText}>Select Medicine</Text>
          <ScrollView horizontal={true} showsHorizontalScrollIndicator={true}>
            {medsArray &&
              medsArray.map(medItem => {
                return <RenderMeds item={medItem}></RenderMeds>;
              })}
          </ScrollView>
        </View>
        <View style={styles.cnView}>
          {mycaretakers.map(item => {
            return <Renderitem item={item}></Renderitem>;
          })}

          <Button
            disabled={send_to === ''}
            id="sendImage"
            onPress={SendImage}
            title="Send"
            buttonStyle={styles.button}
            containerStyle={styles.buttonConatiner}></Button>
        </View>
      </ScrollView>
    </View>
  );
};

export default SendImageToCaretaker;
