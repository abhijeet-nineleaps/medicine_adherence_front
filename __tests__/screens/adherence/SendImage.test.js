import React from 'react';
import {create} from 'react-test-renderer';
import SendImageToCaretaker from '../../../src/screens/adherence/SendImageToCaretaker';
import Enzyme from 'enzyme';
import {shallow} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
Enzyme.configure({adapter: new Adapter()});
jest.mock('react-native-share', () => ({
  default: jest.fn(),
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useFocusEffect: jest.fn(),
  useNavigation: () => ({goBack: jest.fn()}),
  useRoute: () => ({
    params: {
      image_uri: {},
    },
  }),
}));
describe('Send Image Screen', () => {
  it('renders correctly', () => {
    const tree = create(<SendImageToCaretaker navigation={undefined} />);
    expect(tree).toMatchSnapshot();
  });
  it('test open save button', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<SendImageToCaretaker SendImage={mockFn} />);
    wrapper.find('#sendImage').simulate('press');
  });
  it('test open save button', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<SendImageToCaretaker shareFnc={mockFn} />);
    wrapper.find('#share').simulate('press');
  });
});
