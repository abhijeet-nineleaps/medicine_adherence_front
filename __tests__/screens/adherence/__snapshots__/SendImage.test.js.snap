// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send Image Screen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "height": "100%",
      "padding": 20,
    }
  }
>
  <View
    collapsable={false}
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    pointerEvents="box-none"
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
    testID="toastAnimatedContainer"
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#FFF",
          "borderLeftColor": "#69C779",
          "borderLeftWidth": 5,
          "borderRadius": 6,
          "elevation": 2,
          "flexDirection": "row",
          "height": 60,
          "opacity": 1,
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 6,
          "width": 340,
        }
      }
      testID="toastTouchableContainer"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "flex": 1,
              "justifyContent": "center",
              "paddingHorizontal": 25,
            },
            undefined,
          ]
        }
        testID="toastContentContainer"
      />
    </View>
  </View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    style={
      Object {
        "alignItems": "center",
      }
    }
    transparent={true}
    visible={false}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "margin": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#242424",
          "fontWeight": "900",
        }
      }
    >
      Image
    </Text>
    <View
      style={
        Array [
          Object {
            "borderRadius": 3,
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 3,
          },
          undefined,
          false,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Share
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Image
    source={
      Object {
        "uri": Object {
          "image_uri": Object {},
        },
      }
    }
    style={
      Object {
        "borderRadius": 20,
        "height": "60%",
        "width": "100%",
      }
    }
  />
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#242424",
              "marginLeft": 8,
            }
          }
        >
          Select Medicine
        </Text>
        <RCTScrollView
          horizontal={true}
          showsHorizontalScrollIndicator={true}
        >
          <View />
        </RCTScrollView>
      </View>
      <View
        style={
          Object {
            "padding": 50,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 3,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 3,
              },
              Object {
                "marginTop": 25,
              },
              false,
            ]
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "busy": false,
                "disabled": true,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "hsl(208, 8%, 90%)",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "hsl(208, 8%, 63%)",
                    "fontSize": 18,
                    "paddingVertical": 1,
                    "textAlign": "center",
                  }
                }
              >
                Send
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
